src/java/org/apache/cassandra/net/InboundMessageHandlers.java

```diff
@@ -128,7 +132,25 @@ public InboundMessageHandlers(InetAddressAndPort self,
         largeCallbacks  = makeMessageCallbacks(peer, largeCounters,  globalMetricCallbacks, messageConsumer);
         legacyCallbacks = makeMessageCallbacks(peer, legacyCounters, globalMetricCallbacks, messageConsumer);
 
-        metrics = new InternodeInboundMetrics(peer, this);
+        if (CUSTOM_INTERNODE_INBOUND_METRICS_PROVIDER_PROPERTY.isPresent())
```

Comment:
this is error prone.

Instead, there should be a factory method in `InternodeInboundMetrics`, and the `InternodeInboundMetrics` constructor should be private. That way, we can ensure the metrics provider is used every time we construct `InternodeInboundMetrics`